src/topology/homotopy/homotopy_group.lean

```diff
@@ -0,0 +1,209 @@
+/-
+Copyright (c) 2021 Roberto Alvarez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Roberto Alvarez
+-/
+
+import topology.homotopy.basic
+import topology.path_connected
+import topology.unit_interval
+import algebraic_topology.fundamental_groupoid.fundamental_group
```

Comment:
Defining concatenation by coordinate is turning out to be troublesome, continuity in particular, also I haven't been able to define it in terms of path concatenation so that the code in [fundamental_groupoid/basic](https://github.com/leanprover-community/mathlib/blob/master/src/algebraic_topology/fundamental_groupoid/basic.lean) can be reused, so there's much code duplication.

I think using the currying homeomorphism is a better idea, thanks for the pointer!

Reply:
If we adopt the currying approach, I think the first step is to construct `gen_loop (n+1) x ≃ path const const`, where `const : gen_loop n x`, with the compact-open topology on `gen_loop n x` as a subspace of `C(I^n, X)`. (You get one such equivalence for each of the (n+1) coordinates.) Then, we would show that the type [path.homotopy](https://leanprover-community.github.io/mathlib_docs/topology/homotopy/path.html#path.homotopy) between two `path const const` is in bijection with `homotopy_rel` boundary between the two corresponding `gen_loop (n+1) x`. That would allow us to take full advantage of the existing fundamental groupoid constructions and theorems.

Reply:
Iterating `gen_loop (n+1) x ≃ path const const` + some continuity argument should allow us to show that `gen_loop n x` is homeomorphic to the `n`th iterated loop space `X`, so I would not object if you change the name to `loop_space` :)

Reply:
The iterated loop space homeomorphism was also suggested by Prof. Bartholdi, however unless it makes an earlier goal easier it'll remain later in the queue.

The priority goes like this: group/abelian instace > paths induce homomorphisms functorially > group homomorphisms instances > iterated loop space homeomorphism